lib/amazon/index.js

```diff
@@ -7,6 +7,21 @@ var apiUrl = {
 	production: 'https://appstore-sdk.amazon.com/version/1.0/verifyReceiptId/developer/'
 };
 
+function parseResult(resultString) {
+	var result = JSON.parse(resultString);
+
+	var purchaseDate = result.purchaseDate ? parseInt(result.purchaseDate, 10) : null;
+	var expirationDate = result.cancelDate ? parseInt(result.cancelDate, 10) : null;
```

Comment:
Instead of a falling back to null, we should use renewal date:

```
var expirationDate = result.cancelDate ? parseInt(result.cancelDate, 10)  : parseInt(result.renewalDate, 10)
```

Why?

When a cancellation date is not defined, a customer has entitlement up until their renewal date. If they don't renew, then this date is still valid as it represents their permitted access. Cancel date should always be a first-try, as you are doing, because when a customer voluntarily cancels or they are cancelled due to non-payment, that date will become populated. Otherwise, the renewal date can be used as it implies how long someone has access until they need to renew.

If both the cancel date and the renewal date is null, then providing null back is fine.

Reply:
Hmm not sure why I used the cancel date. I meant to use the `renewalDate`.

On iOS the cancel date and expiration date are different. For example if I subscribe today for 1 month. Then cancel 2 weeks later, expiration date is still 2 weeks away and the user should still have access until the expiration date, not the cancel date. Essentially the cancel date is the date that the auto-renew was turned off and not the expiration date. On Amazon, this seems to be different.

On Amazon, if I cancel 2 weeks after the purchase, should that be the expiration date and should user access be revoked? Does Amazon refund the user for the remaining 2 weeks? If that is the case, then using the `cancelDate` first and then fall back to `renwalDate` seems to be right.

However, if that is not the case, maybe we can check for renewalDate first. If that is not available then the user might have cancelled or not paid. Then can look up the cancelDate, which should be populated. However, here I am assuming that the renewalDate is set to null, if user doesn't pay.

https://developer.amazon.com/docs/in-app-purchasing/iap-rvs-for-android-apps.html
`
 If the receipt is for an expired subscription the cancelDate would reflect the date the subscription expired, or the date it was canceled by Amazon customer support.
`

What do you think? Maybe I'm missing something.

Reply:
Ok, sounds good to me.